sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/DotNetSensor.java

```diff
@@ -73,16 +72,24 @@ public void execute(SensorContext context) {
   }
 
   private boolean shouldExecuteOnProject(FileSystem fs) {
-    if (!filesToAnalyze(fs).iterator().hasNext()) {
-      LOG.debug("No files to analyze. Skip Sensor.");
+    boolean hasMainFiles = hasFilesOfType(fs, Type.MAIN);
+    boolean hasTestFiles = hasFilesOfType(fs, Type.TEST);
+
+    if (!hasMainFiles) {
+      if (hasTestFiles) {
+        String noResultsMessage = "Your SonarQube/SonarCloud project will not have results for " + pluginMetadata.languageKey() + " files.";
+        LOG.warn("This solution contains only TEST files. Because there are no MAIN files, this sensor will be skipped. " + noResultsMessage);
+      } else {
+        LOG.debug("No files to analyze. Skip Sensor.");
+      }
```

Comment:
I'd do two changes:
- Make conditions positive.
- I think that 'hasFilesOfType' is expensive under the hood. So call it only when needed.
```suggestion
    if (hasFilesOfType(fs, Type.MAIN)) {
      return true;
    } else if (hasFilesOfType(fs, Type.TEST)) {
        String noResultsMessage = "Your SonarQube/SonarCloud project will not have results for " + pluginMetadata.languageKey() + " files.";
        LOG.warn("This solution contains only TEST files. Because there are no MAIN files, this sensor will be skipped. " + noResultsMessage);
    } else {
      LOG.debug("No files to analyze. Skip Sensor.");
    }
    return false;
```